### PR TITLE
resolved conflicts added new features

### PR DESCRIPTION
**Pull request by DrogoNevets**

### CODE REVIEW

File: src/controllers/ofsted.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    set isASP(val: boolean) {
+        this._hasASP = val;
+    }
+}
+
+@Controller('ofsted')
+class OfstedController {
+    private IDSR_ID: string = 'e726a43f-d1d6-456d-8d70-ea7396a97394';
+    private GIAS_SVC_ID: string = '2354cb2e-f559-4bf4-9981-4f6c6890aa5e'
+    private ASP_SVC_ID: string = 'df2ae7f3-917a-4489-8a62-8b9b536a71cc';
```
Comment (joseluisgraa) [inline src/controllers/ofsted.ts:57]: These service IDs should come from env vars if they can change between environments

File: src/controllers/ofsted.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                    reportUser.isGIAS = true;
+                } else {
+                    reportUser.isGIAS = false;
+                }
+
+                if (aspUser) {
+                    reportUser.isASP = true;
+                } else {
+                    reportUser.isASP = false;
+                }
```
Comment (joseluisgraa) [inline src/controllers/ofsted.ts:93]: I think you could do this in two lines:
reportUser.isGIAS = !!giasUser;
reportUser.isASP = !!aspUser;

File: src/services/databaseOrg.ts [hunk 3]
```diff
@@ -33,13 +43,31 @@ class MSSQLDataBaseOrg {
...
+    let Id = userId === null ? userId : `'${userId}'`
+    let sql = `SELECT distinct top 1 uo.numeric_identifier, uo.text_identifier, o.UKPRN, o.URN, o.UID, o.legacyId from user_organisation as uo join organisation as o on 
                 id=organisation_id where uo.user_id = ${Id} and uo.organisation_id =${orgId};`;
-      let result = await this.pool.request().query(sql);
-      return Promise.resolve(result.recordset);
+    let result = await this.pool.request().query(sql);
+    return Promise.resolve(result.recordset);
+  }
+
+  async findUserServiceByService(serviceId: string): Promise<Array<any>> {
```
Comment (joseluisgraa) [inline src/services/databaseOrg.ts:55]: maybe call this findUserServiceByServiceId?

File: src/services/jobs.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
+import { request } from "https";
+
+import tokenSvc, { ITokenOpts } from './token';
+import redisSvc from './redis';
+import sleep from '../utils/sleep';
+
+class JobsService {
+    private async generateToken(): Promise<string> {
+        let tokenConfig: ITokenOpts = {
+            url: "https://signin-pr-dirlnx-as.azurewebsites.net",
```
Comment (joseluisgraa) [inline src/services/jobs.ts:10]: Shouldn't this URL be environment specific?

File: src/views/includes/team-sidebar.ejs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    <em>Developer</em>
+</div>
+
+<div class="govuk-inset-text">
+    Arshed Mustafa<br />
+    <em>Developer</em>
+</div>
+
+<div class="govuk-inset-text">
+    ???<br />
```
Comment (joseluisgraa) [inline src/views/includes/team-sidebar.ejs:21]: Missing some names here ;)

File: src/controllers/ofsted.ts (hunk 2 above)
Comment (DrogoNevets) [reply src/controllers/ofsted.ts:93]: considered it, questioned readability. Thoughts?

File: src/services/jobs.ts (hunk 4 above)
Comment (DrogoNevets) [reply src/services/jobs.ts:10]: ah yes - will change this too

File: src/controllers/ofsted.ts (hunk 2 above)
Comment (joseluisgraa) [reply src/controllers/ofsted.ts:93]: As long as we know what !! is doing I think it's fine, but that's only my opinion!

File: src/services/redis.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    async inactiveJobsQueuing(): Promise<boolean> {
+        this.connect();
+
+        await this.getInactiveJobCount();
+
+        this.disconnect();
+
+        let countDifference = (a: number, b: number): number => b - a,
+            counts: Array<number> = this.inactiveJobCounts,
+            restart: boolean = ((counts.length >= 3 && countDifference(counts[0], counts[1]) > 10 && countDifference(counts[1], counts[2]) > 10 && counts[2] !== 0) ||
```
Comment (joseluisgraa) [inline src/services/redis.ts:36]: not sure what's going on here

File: src/services/redis.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                (counts[0] === counts[1] && counts[1] === counts[2] && counts[2] !== 0) ||
+                (counts[0] > 150 && counts[1] > 150 && counts[2] > 150));
+
+        return Promise.resolve(restart);
+    }
+
+    async getInactiveJobCount(): Promise<number> {
+        let count;
+
+        while (this.inactiveJobCounts.length > 2) {
```
Comment (joseluisgraa) [inline src/services/redis.ts:46]: not sure what we are doing here, what is the content of inactiveJobCounts exactly? why are we removing items from the array?

File: src/services/jobs.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            });
+
+            req.write(DATA);
+            req.end();
+        });
+    }
+
+    async start(): Promise<void> {
+        return new Promise(async (resolve, reject) => {
+            const DATA = JSON.stringify({ status: 'stopped' }),
```
Comment (joseluisgraa) [inline src/services/jobs.ts:82]: I guess status here should be 'started'

File: src/services/databaseOrg.ts [hunk 9]
```diff
@@ -33,13 +43,31 @@ class MSSQLDataBaseOrg {
...
+    let sql = `SELECT * FROM [dbo].[user_services] WHERE service_id='${serviceId}'`;
+
+    return this.query(sql);
+  }
+
+  async usersHaveAccessTo(users: Array<string>, svcId: string): Promise<Array<any>> {
```
Comment (joseluisgraa) [inline src/services/databaseOrg.ts:61]: perhaps I would call this findUsersWithAccessTo but again personal preference.

File: src/controllers/jobs.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+            },
+            (err: Error, html: string) => {
+                res
+                    .status(200)
+                    .send(html)
+                    .end();
+            }
+        );
+
+        return;
```
Comment (joseluisgraa) [inline src/controllers/jobs.ts:35]: shouldn't this need to be returning a promise? or it is fine as is?

File: src/services/redis.ts (hunk 6 above)
Comment (DrogoNevets) [reply src/services/redis.ts:36]: basically a massive IF statement of the conditions on which we trigger a bounce of the jobs queue

File: src/services/redis.ts (hunk 7 above)
Comment (DrogoNevets) [reply src/services/redis.ts:46]: so as to only store the most recent 3 counts, we dont need a complete history just a recent one to see if the jobs count is increasing or decreasing

File: src/services/jobs.ts (hunk 8 above)
Comment (DrogoNevets) [reply src/services/jobs.ts:82]: good spot

File: src/services/databaseOrg.ts (hunk 9 above)
Comment (DrogoNevets) [reply src/services/databaseOrg.ts:61]: I likey

File: src/controllers/jobs.ts (hunk 10 above)
Comment (DrogoNevets) [reply src/controllers/jobs.ts:35]: by being a `async` method it always returns a Proimise, but it a bit ambiguous I agree - will be explicit

File: src/services/redis.ts (hunk 7 above)
Comment (joseluisgraa) [reply src/services/redis.ts:47]: can we have a comment here saying that? something like "we only need to keep 3 counts in the history" or similar?

File: src/services/redis.ts (hunk 6 above)
Comment (joseluisgraa) [reply src/services/redis.ts:37]: I see you have added a comment but it would be great for someone new looking at it to see a comment with the actual business logic behind the condition. Something like "restart if difference between last counts is greater than X or if all inactive jobs are same number or whatever", in plain English
